components/form.tsx

```diff
@@ -2,14 +2,90 @@
 
 import Link from 'next/link';
 import Image from 'next/image';
+import { useForm } from 'react-hook-form';
+import type { SubmitHandler } from 'react-hook-form';
+import { z } from 'zod';
+import { zodResolver } from '@hookform/resolvers/zod';
 
-import { Input, Button } from '@/components';
+import { Input, Button, Spinner, Success } from '@/components';
+
+interface FormFields {
+  email: string;
+  username: string;
+  password: string;
+  confirmPassword: string;
+}
+
+export const UserSchema = z
+  .object({
+    email: z
+      .string()
+      .min(1, 'Email is required.')
+      .email('Please enter a valid email address.')
+      .regex(
+        /^[a-zA-Z0-9._%+-]+@(gmail\.com|hotmail\.com|lightit\.io)$/,
+        'Email must be from one of the allowed domains: gmail.com, hotmail.com, or lightit.io.',
+      )
+      .trim(),
+    username: z
+      .string({ required_error: 'Username is required.' })
+      .min(1, 'Username is required.')
+      .toLowerCase()
+      .regex(
+        /^[a-z]+$/,
+        'Username can only contain lowercase letters, no spaces, numbers, or special characters.',
+      )
+      .min(3, 'Username must be at least 3 characters long.')
+      .max(32, 'Username must be 32 characters or less.')
+      .trim(),
+    password: z
+      .string({ required_error: 'Password is required.' })
+      .min(1, 'Password is required.')
+      .regex(
+        /^(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]+$/,
+        'Password must contain at least one uppercase letter, one lowercase letter, and one special character.',
+      )
+      .min(8, 'Password must be at least 8 characters long.')
+      .max(64, 'Password must be 64 characters or less.'),
+    confirmPassword: z.string(),
+  })
+  .refine((data) => data.password === data.confirmPassword, {
+    path: ['confirmPassword'],
+    message: 'Passwords do not match.',
+  });
```

Comment:
You can use `z.infer` for the form values:

```suggestion

export const userSchema = z
  .object({
    email: z
      .string()
      .min(1, 'Email is required.')
      .email('Please enter a valid email address.')
      .regex(
        /^[a-zA-Z0-9._%+-]+@(gmail\.com|hotmail\.com|lightit\.io)$/,
        'Email must be from one of the allowed domains: gmail.com, hotmail.com, or lightit.io.',
      )
      .trim(),
    username: z
      .string({ required_error: 'Username is required.' })
      .min(1, 'Username is required.')
      .toLowerCase()
      .regex(
        /^[a-z]+$/,
        'Username can only contain lowercase letters, no spaces, numbers, or special characters.',
      )
      .min(3, 'Username must be at least 3 characters long.')
      .max(32, 'Username must be 32 characters or less.')
      .trim(),
    password: z
      .string({ required_error: 'Password is required.' })
      .min(1, 'Password is required.')
      .regex(
        /^(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*])[A-Za-z\d!@#$%^&*]+$/,
        'Password must contain at least one uppercase letter, one lowercase letter, and one special character.',
      )
      .min(8, 'Password must be at least 8 characters long.')
      .max(64, 'Password must be 64 characters or less.'),
    confirmPassword: z.string(),
  })
  .refine((data) => data.password === data.confirmPassword, {
    path: ['confirmPassword'],
    message: 'Passwords do not match.',
  });
  
  type UserFormValuesType = z.infer<typeof userSchema>;
```

Reply:
Also consider adding the schema in a separate file as `userSchema.ts` inside utils as a good practice 🙂

Reply:
Thanks for the advice! I’ll also move the schema to a separate file as suggested.